crates/core_simd/src/vector.rs

```diff
@@ -238,19 +240,42 @@ where
         idxs: Simd<usize, LANES>,
     ) {
         let enable: Mask<isize, LANES> = enable & idxs.lanes_lt(Simd::splat(slice.len()));
-        // SAFETY: We have masked-off out-of-bounds lanes.
-        unsafe { self.scatter_select_unchecked(slice, enable, idxs) }
+        // SAFETY: Out-of-bounds lanes have been masked, and the ptr is immediately materialized.
+        unsafe { self.scatter_select_unchecked(slice.as_mut_ptr(), enable, idxs) }
     }
 
-    /// Writes the values in a SIMD vector to multiple potentially discontiguous indices in `slice`.
+    /// Writes the values in a SIMD vector to multiple potentially discontiguous locations,
+    /// offset from the `ptr` provided by indices, assuming it is a base ptr to `[T]`.
     /// The mask `enable`s all `true` lanes and disables all `false` lanes.
     /// If two enabled lanes in the scattered vector would write to the same index,
     /// only the last lane is guaranteed to actually be written.
     ///
     /// # Safety
     ///
-    /// Calling this function with an enabled out-of-bounds index is *[undefined behavior]*,
-    /// and may lead to memory corruption.
+    /// - Calling this function with any `*mut T` that is not a base pointer to `[T]` or a type
+    ///   with equivalent layout (such as arrays or SIMD types) results in *[undefined behavior]*.
+    /// - Calling this function with an `enable`d out-of-bounds index results in *[undefined behavior]*.
+    /// - If using `&mut [T]` or similar, such as `Vec<T>` or `&mut [T; N]`, it is safest to
+    /// construct the bounds mask first if applicable and then use `fn as_mut_ptr(&mut self)`
+    /// to obtain *mut T` immediately as part of this method call, with nothing else between,
+    /// as even deriving `&[T]` from by calling `fn len` creates an `&[T]` which,
+    /// as an immutable borrow derived from a mutable (and therefore unique) borrow,
+    /// interacts in a way that invalidates then-live raw `*mut T`s according to the
+    /// rules of "Stacked Borrows", which can lead to *[undefined behavior]*.
```

Comment:
I'm still not really convinced this is currently the case, rather than a proposed rule...

Reply:
Right, stacked borrows is a proposed aliasing model, rather than an accepted one. There are things in the language that it does not properly model, and so it probably cannot be the final one. I think we can probably be a lot more vague about what is UB here, though.

Also, I'd expect it to have rules derived from a combination of the rules of https://doc.rust-lang.org/std/ptr/fn.write.html (which heavily leans on https://doc.rust-lang.org/std/ptr/index.html) and the rules of https://doc.rust-lang.org/std/primitive.pointer.html#method.offset

Reply:
As there is no rule that is known to be true,
there is also no rule that is certain to be false.
therefore, in the interests of users not having to rewrite their code
we should advise them to duck the maybe-brick that may or may not collide with their skull.

It is true the user's cat may still be alive if we open the box,
or perhaps the user will be a butterfly.

Reply:
Ah yes, I need to specify the aligned/non-null clause.

Reply:
I think it is even more confusing to bring up rules that don't currently exist.  This is certainly more strict than LLVM's rules so I don't think it's right to portray it as the current state of rust.  In my opinion scatter is just one of many, many APIs that use pointers and isn't special enough to warrant any explanation at all.  I think the simplest and most correct way to document this is to specify that it follows all the same requirements as `ptr::write` (whatever those may be at any particular point in time).

Reply:
>  This is certainly more strict than LLVM's rules 
Correct.

But one should not backpropagate from LLVM's rules, which are themselves ambiguous, poorly specified, and likely to become more strict in the future, to the Rust programming language, for the same reason one should not backpropagate rules from x86 to the C programming language. C is certainly more restrictive in many ways than x86 is, often.

I also think that it's likely that Rust will have a more lax model in the future.

But as far as I have observed, the users of unsafe code complain most about the prospect of having to rewrite that code based on future model changes. Specifying this in the way least likely to be invalidated seems soundest, even if it is arguably "most incorrect", as it takes the upper bound on constraints, rather than an average.

Reply:
It's problematic for vector scatters/gathers to have stricter rules than then underlying operations. It makes it harder to verify if vectorized code is correct, and it would mean that things that analyze UB (which includes, but is not limited to miri, CC @RalfJung) have to special-case this, or have undetectable UB (which is undesirable).

That is, assuming these are intended to be language-level UB rather than library-level UB (which seems to be the case)

Reply:
It's not a stricter rule than the underlying operations, it's that the most obvious ways to set up the environment **before** you call the function have a gotcha if you do them out of order. This documentation actually has nothing to do with the function itself, but has everything to do with me trying to address that gotcha in the documentation. I am not interested, however, in entertaining an argument here over this. I am withdrawing this PR.

Reply:
FWIW I agree that ideally the safety contract here would be stated in the same terms as the `core::ptr` functions. If that seems insufficiently clear, we should probably clarify those rules.

> It's not a stricter rule than the underlying operations, it's that the most obvious ways to set up the environment before you call the function have a gotcha if you do them out of order. 

Looks like it's probably not worth further discussing this PR, but for the future -- I think what I am missing here is what that most obvious way looks like. It would help to have a concrete example. If there is something about SIMD code that makes this more likely to be a problem than scalar code, I would understand better why you are suggesting SIMD should document this more carefully than the scalar functions we have.

Reply:
You yourself admonished me:
![image](https://user-images.githubusercontent.com/46493976/145728256-b93750a0-8f89-4ffc-b1c1-ba81f03d0652.png)

Reply:
So in the usual way of calling this raw-ptr-taking function, there is a call to `len` involved in a way that is problematic?

Basically I don't quite understand what *concrete* problem these comments are referring to. The block of code "must be exactly this way" to make sure things work out, but what would the other way(s) even be? It is not possible to reorder these lines since each uses a result just computed in the previous line.

Reply:
I think that Jubilee is referring to the general fact that Stacked Borrows makes *all* interaction with a mutable slice (including this example here) into a general footgun if the user makes the mutable pointer first and then gets the length of the slice after that. Since people generally consider slices to be "a pointer and a length", they are often inclined to write the code similarly:
```rust
let p = my_slice.as_mut_ptr();
let len = my_slice.len();
```
and now they've blown off their foot because `p` is invalid.

It's nothing particular to this function, and nothing to do with simd. Stacked Borrows just has a footgun for all mutable slices.

Reply:
Yes, if a caller starts from a Vec (which is to be expected) then they may call `scatter_select_unchecked` in a way that would mimic the following pattern:
```rust
    let mut v = vec![4, 5, 6, 7];
    let mut x = v.as_mut_ptr();
    unsafe {
        x = x.offset(1);
        if x < v.as_mut_ptr().add(v.len()) {
            x.write(0);
        }
    }
```
Because why wouldn't they?

Reply:
Lokathor is correct, yes. I just think it is important to reiterate the warning in appropriately flamingly large text, but if I say that, then everyone complains at me that Miri doesn't issue a warning on that code.

The oddity of the (old) internal comments being distracting in the presence of the (new) external documentation is merely a relic of me focusing on trying to be agonizingly clear in the external documentation.

Reply:
Okay, I see -- thanks.

I think people were mostly complaining that *because* this is such a general issue, it seems odd to have it explained in user-visible documentation in this particular function. If we want to warn people in the docs about these patterns which Stacked Borrows disallows (but which other future aliasing models might allow), we should probably do that more centrally and more consistently.

Reply:
Sure. But my primary concern isn't really if all of `core::ptr` is properly documented, it's that one method, `core::simd::Simd::scatter_select_unchecked`, is properly documented, one likely to be used by people who **don't** want to read a ton of general documentation.

If that is truly the concern, then we could have figured out where to move the docs from this method (and replace it with links to that location) after this PR.

Reply:
> If that is truly the concern, then we could have figured out where to move the docs from this method (and replace it with links to that location) after this PR.

That seems to be assuming the consensus that we want to have Stacked Borrows specific docs in the official doc comments at all, which AFAIK is an open question.

Reply:
And so we come full circle to "But as far as I have observed, the users of unsafe code complain most about the prospect of having to rewrite that code based on future model changes."
 ¯\_(ツ)_/¯